### PR TITLE
feat: add context manager for register_type

### LINKED ISSUE #466: Feature: add context manager to temporarily change the widget associated with a type
@Czaki had a good request:  a context manager like the following:

```python
with type_registered(MyType, widget_type=MyWdgClass):
    ...
``` 

that would restore any previously existing association with `MyType` when the context is exited

### PR DESCRIPTION
**Pull request by tlambert03**
closes #466 

@Czaki ... if you have a moment to review, it would be appreciated!

### CODE REVIEW

File: magicgui/type_map.py [hunk 1]
```diff
@@ -369,6 +380,58 @@ def _deco(type_):
...
+    null = object()
+
+    prev_return_callback = _RETURN_CALLBACKS.get(_type_, null)
+    prev_type_def = _TYPE_DEFS.get(_type_, null)
+    try:
+        register_type(
+            type_, widget_type=widget_type, return_callback=return_callback, **options
+        )
+        yield
+    finally:
```
Comment (Czaki) [inline magicgui/type_map.py:424]: what did you think about add here check if values under `_RETURN_CALLBACKS[_type_]` and `_TYPE_DEFS[_type_]`  are such as expected and if not (someone calls `register_type` for such type inside context manager) then print warning message?

File: magicgui/type_map.py [hunk 2]
```diff
@@ -369,6 +380,58 @@ def _deco(type_):
...
+    tw.resolve()
+    _type_ = tw.outer_type_
+    null = object()
+
+    prev_return_callback = _RETURN_CALLBACKS.get(_type_, null)
+    prev_type_def = _TYPE_DEFS.get(_type_, null)
+    try:
+        register_type(
+            type_, widget_type=widget_type, return_callback=return_callback, **options
+        )
```
Comment (Czaki) [inline magicgui/type_map.py:422]: having this under `try` clause is to keep the consistency of the data structure. If yes, then maybe this should be done in `register_type` itself?

File: magicgui/type_map.py [hunk 3]
```diff
@@ -369,6 +380,58 @@ def _deco(type_):
...
+        yield
+    finally:
+        if prev_return_callback is null:
+            _RETURN_CALLBACKS.pop(_type_, None)
+        else:
+            _RETURN_CALLBACKS[_type_] = prev_return_callback  # type: ignore
+        if prev_type_def is null:
+            _TYPE_DEFS.pop(_type_, None)
+        else:
+            _TYPE_DEFS[_type_] = prev_type_def  # type: ignore
```
Comment (Czaki) [inline magicgui/type_map.py:432]: looking at test and coverage messages, it looks like something is wrong. Or are default widgets stored outside `_TYPE_DEFS` dict?

Comment (tlambert03) [reply magicgui/type_map.py:432]: default widgets are stored outside of `_TYPE_DEFS` (for better or worse)

File: magicgui/type_map.py (hunk 2 above)
Comment (tlambert03) [reply magicgui/type_map.py:422]: i agree, bit odd in the try block there.  gonna just move it above.

File: magicgui/type_map.py (hunk 1 above)
Comment (tlambert03) [reply magicgui/type_map.py:424]: added, thanks

**Comment by tlambert03** [comment]
many thanks for the review.  updated

File: magicgui/type_map.py [hunk 4]
```diff
@@ -369,6 +381,60 @@ def _deco(type_):
...
+    _type_ = register_type(
+        type_, widget_type=widget_type, return_callback=return_callback, **options
+    )
+    new_type_def: Optional[WidgetTuple] = _TYPE_DEFS.get(_type_, None)
+    try:
+        yield
+    finally:
+
+        if return_callback is not None:
+            _RETURN_CALLBACKS[_type_].remove(return_callback)
```
Comment (Czaki) [inline magicgui/type_map.py:427]: in the previous implementation, you restore also value of `_RETURN_CALLBACKS` here you only remove some value. 

What happen if someone use `register_type` with both `widget_type` and `return_callback` inside? 

Maybe also callback should be removed only if is not present in `_RETURN_CALLBACKS[_type_]` before the context manager usage?

Comment (tlambert03) [reply magicgui/type_map.py:430]: > Maybe also callback should be removed only if is not present in _RETURN_CALLBACKS[_type_] before the context manager usage?

sounds good, added

**Pull request merged**